src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenStringConcat.cs

```diff
@@ -2104,5 +2105,1072 @@ .maxstack  5
+                  IL_0000:  ldarg.0
+                  IL_0001:  ldarga.s   V_1
+                  IL_0003:  call       "string char.ToString()"
+                  IL_0008:  ldarg.0
+                  IL_0009:  call       "string string.Concat(string, string, string)"
+                  IL_000e:  ret
+                }
+                """);
+            verifier.VerifyIL("Test.M3", """
+                {
+                  // Code size       15 (0xf)
+                  .maxstack  3
+                  IL_0000:  ldarg.0
+                  IL_0001:  ldarg.0
+                  IL_0002:  ldarga.s   V_1
+                  IL_0004:  call       "string char.ToString()"
+                  IL_0009:  call       "string string.Concat(string, string, string)"
+                  IL_000e:  ret
+                }
+                """);
+            verifier.VerifyIL("Test.M4", """
+                {
+                  // Code size       21 (0x15)
+                  .maxstack  3
+                  IL_0000:  ldarga.s   V_1
+                  IL_0002:  call       "string char.ToString()"
+                  IL_0007:  ldarg.0
+                  IL_0008:  ldarga.s   V_1
+                  IL_000a:  call       "string char.ToString()"
+                  IL_000f:  call       "string string.Concat(string, string, string)"
+                  IL_0014:  ret
+                }
+                """);
+        }
+
+        [Theory, WorkItem("https://github.com/dotnet/roslyn/issues/66827")]
+        [InlineData((int)WellKnownMember.System_String__Concat_ReadOnlySpanReadOnlySpan)]
+        [InlineData((int)WellKnownMember.System_String__Concat_ReadOnlySpanReadOnlySpanReadOnlySpan)]
+        public void ConcatThree_ReadOnlySpan_MissingSpanConcat(int spanConcatMember)
+        {
+            var source = """
+                using System;
+
+                public class Test
+                {
+                    static void Main()
+                    {
+                        var s = "s";
+                        var c = 'c';
+                        Console.Write(M1(s, c));
+                        Console.Write(M2(s, c));
+                        Console.Write(M3(s, c));
+                        Console.Write(M4(s, c));
+                    }
+
+                    static string M1(string s, char c) => c + s + s;
+                    static string M2(string s, char c) => s + c + s;
+                    static string M3(string s, char c) => s + s + c;
+                    static string M4(string s, char c) => c + s + c;
+                }
+                """;
+
+            var comp = CreateCompilation(source, options: TestOptions.ReleaseExe, targetFramework: TargetFramework.Net80);
+            comp.MakeMemberMissing((WellKnownMember)spanConcatMember);
+
+            // Just verify that we can still run this and get expected output.
+            // This is not something that can be seen in real-life scenarios, so don't care about precise IL we generate
```

Comment:
Isn't the expected output and IL the same as in the test above? Cannot it be merged into one test?

Reply:
Unfortunatelly not. Let's consider `s + c + s` example, where `s` is a string and `c` is a char. Compiler lowers this left-to-right, meaning, that first `s + c` is lowered and then happens lowering of its result and the final `s`. On the first iteration if we pick up optimized path we get `string.Concat(s, new ReadOnlySpan<char>(in c))`. On the second one we "unwrap" `string.Concat` back, receivig `s`, wrapped into an implicit conversion to a span, and `new ReadOnlySpan<char>(in c)`. Then we wrap the last `s` into a a span and get `string.Concat(s, new ReadOnlySpan<char>(in c), s)`. And now let's see what if only 1 span-based `Concat` overload is missing. If we miss span concat of 2 on first iteration we get `string.Concat(s, c.ToString())`, so when we get to the second one we unwrap 2 string arguments and emit `string.Concat(s, c.ToString(), s)`. However if we are missing span concat of 3 we get `string.Concat(s, new ReadOnlySpan<char>(in c))` on the first iteration, but then we don't unwrap it on the second iteration and get `string.Concat(string.Concat(s, new ReadOnlySpan<char>(in c)), s)`. So yeah, unfortunatelly, it we want to check IL we can only do it separately for each case, which IMO is not worth it since in reality case when only 1 span-based `Concat` is missing is not possible